txpool/txpool.go

```diff
@@ -823,466 +531,508 @@ func (t *TxPool) validateTx(
-		return nil
+// handleEnqueueRequest is invoked when a new transaction is received
+// (result of a successful addTx() call) or a demoted transaction
+// is re-entering the pool.
+//
+// A transaction handled within this request can either be
+// dropped or enqueued, potentially signaling promotion in the latter case.
+func (p *TxPool) handleEnqueueRequest(req enqueueRequest) {
+	tx := req.tx
+	addr := req.tx.From
+
+	// fetch account
+	account := p.accounts.from(addr)
+
+	// enqueue tx
+	if err := account.enqueue(tx, req.demoted); err != nil {
+		p.logger.Error("enqueue request", "err", err)
+		return
 	}
 
-	// reject remote tx with lower gasPrice
-	// than the min gasPrice tx currently present in remoteTxns
-	if !isLocal && t.Underpriced(tx) {
-		return ErrUnderpriced
-	}
+	p.logger.Debug("enqueue request", "hash", tx.Hash.String())
 
-	// try to allocate space
-	overflow := t.gauge.height + txSlots - t.gauge.limit
-	dropped, success := t.Discard(overflow, isLocal)
-	if !isLocal && !success {
-		return ErrTxPoolOverflow
-	}
+	// update lookup
+	p.index.add(tx)
 
-	// clear dropped txs and readjust gauge
-	for _, tx := range dropped {
-		wrapper := t.lockAccountQueue(tx.From, true)
-		wrapper.accountQueue.Remove(tx.Hash)
-		wrapper.unlock()
-
-		t.pendingQueue.Delete(tx)
-		t.gauge.height -= slotsRequired(tx)
+	// demoted transactions never decrease the gauge
+	if !req.demoted {
+		p.gauge.increase(slotsRequired(tx))
 	}
-	t.gauge.height += txSlots
 
-	t.metrics.PendingTxs.Set(float64(t.pendingQueue.Length()))
-	return nil
+	if tx.Nonce <= account.getNonce() {
+		// account queue is ready for promotion:
+		// 	1. New tx is matching nonce expected
+		// 	2. Demoted tx is eligible for promotion
+		p.promoteReqCh <- promoteRequest{account: addr} // BLOCKING
+	}
 }
 
-// txHeapWrapper is a wrapper object for account based transactions
-type txHeapWrapper struct {
-	// txs is the actual min heap (nonce ordered) for account transactions
-	txs txHeap
+// handlePromoteRequest handles moving promotable transactions
+// of some account from enqueued to promoted. Can only be
+// invoked by handleEnqueueRequest or resetAccount.
+func (p *TxPool) handlePromoteRequest(req promoteRequest) {
+	addr := req.account
+	account := p.accounts.from(addr)
 
-	// nextNonce is a field indicating what should be the next
-	// valid nonce for the account transaction
-	nextNonce uint64
+	// promote enqueued txs
+	promoted := account.promote()
+	p.logger.Debug("promote request", "promoted", promoted, "addr", addr.String())
 
-	// Logger used for account-specific tx activity
-	logger hclog.Logger
+	// update metrics
+	p.metrics.PendingTxs.Add(float64(promoted))
 }
 
-// newTxHeapWrapper creates a new account based tx heap
-func newTxHeapWrapper(logger hclog.Logger) *txHeapWrapper {
-	return &txHeapWrapper{
-		logger: logger,
-		txs:    txHeap{},
+// handleGossipTx handles receiving transactions
+// gossiped by the network.
+func (p *TxPool) handleGossipTx(obj interface{}) {
```

Comment:
It might make sense to rename this to addGossipTx, since it's the same kind of wrapper as AddLocalTx (AddTx)

Reply:
Agreed.

[Renamed](https://github.com/0xPolygon/polygon-sdk/pull/287/commits/034e1c7dec054cece5fa24dad0deb98f77404b77)